Libplanet/Blockchain/BlockChain.cs

```diff
@@ -1192,21 +1183,25 @@ public IImmutableSet<TxId> GetStagedTransactionIds()
                             }
                         )
                         .ToImmutableDictionary(t => t.signer, t => t.maxNonce);
-                    ISet<TxId> txIds = Store.IterateStagedTransactionIds()
-                        .Select(Store.GetTransaction<T>)
-                        .Where(tx => maxNonces.TryGetValue(tx.Signer, out long nonce) &&
-                            tx.Nonce <= nonce)
-                        .Select(tx => tx.Id)
-                        .ToImmutableHashSet();
-                    Store.UnstageTransactionIds(txIds);
-                    TipChanged?.Invoke(this, (prevTip, block));
-                    _logger.Debug("Block {blockIndex}: {block} is appended.", block?.Index, block);
                 }
                 finally
                 {
                     _rwlock.ExitWriteLock();
                 }
 
+                ISet<TxId> txIds = StagePolicy.Iterate(this)
+                    .Where(tx => maxNonces.TryGetValue(tx.Signer, out long nonce) &&
+                        tx.Nonce <= nonce)
+                    .Select(tx => tx.Id)
+                    .ToImmutableHashSet();
+                foreach (TxId txId in txIds)
+                {
+                    StagePolicy.Unstage(this, txId);
```

Comment:
Is `StagePolicy.Unstage()` also thread-safe?

Reply:
I made every operation on `IStagePolicy<T>` (protocol-wise) & `VolatileStagePolicy<T>` (implementation-wise) sure they are thread safe: 131ed7ddd7bccc7054cabfa5f06ac3097a7a1796.